### PR TITLE
THREESCALE-10233: Support OpenAPI 3.1 for ActiveDocs

### PR DESCRIPTION
**Pull request by mayorova**
**What this PR does / why we need it**:

Add support for OpenAPI 3.1:
- upgrade to a newer `swagger-ui` version
- use a different validation library for validating OpenAPI - [json_schemer](https://github.com/davishmcclurg/json_schemer)

The new library is required for 3.1 validation, because it's schema depends on JSON Schema version `2020-12`, which is not supported by the current library [json-schema](https://github.com/voxpupuli/json-schema).

We actually use a fork of this library to support `draft7` of JSON Schema: https://github.com/voxpupuli/json-schema/compare/master...3scale:json-schema:master This was required for policies validation. And we eventually might want to switch to `json_schemer` for policies validation.

However, I found out that validation of Swagger 1.2 does not work out of the box with `json_schemer`, so I kept the old validation for it. We don't really "support" Swagger 1.2 anymore, but it kind of still works, so probably we shouldn't break it.

And for Swagger 2.0 I think `json_schemer` could be used, but I decided to keep the current solution.

On the other hand, I decided to move OpenAPI 3.0 to new validation :grimacing:  Just because it's supported out of the box by `json_schemer`, without providing the JSON spec for it.
The drawback is that the validation error messages change in case there are some issues with the spec, but I think it's not a big deal.

**Which issue(s) this PR fixes** 

https://issues.redhat.com/browse/THREESCALE-10233

**Verification steps** 

Create an API spec following OpenAPI v3.1 and make sure it's rendered properly. Example minimal spec:
```
{
  "openapi": "3.1.0",
  "info": {
    "title": "Sample API",
    "description": "Sample API description",
    "version": "1.0.0"
  },
  "servers": [
    {
      "url": "https://echo-api.3scale.net",
      "description": "Echo API"
    }
  ],
  "paths": {
    "/hello": {
      "get": {
        "operationId": "hello",
        "description": "Hello",
        "tags": [
          "greeting"
        ],
        "responses": {
          "200": {
            "description": "OK"
          },
          "400": {
            "description": "Bad request"
          }
        }
      }
    }
  }
}
```

**Special notes for your reviewer**:

There is some strange behavior in the latest versions of the library, apparently caused by this PR: https://github.com/swagger-api/swagger-ui/pull/9762 which first appears in the version [v5.13.0](https://github.com/swagger-api/swagger-ui/releases/tag/v5.13.0). This results in all fields in some requests to be auto-filled with `string` value, which is kind of considered a "default" or "example" value. This is obviously unexpected and very annoying.

swagger-ui-string-initial-value

One possible way of solving it is to define the default values for all fields, but I decided to downgrade to an older version that doesn't have this problem. The main reason being that I consider this behavior wrong and I think it will need to be fixed in `swagger-ui`. I will open an issue for it, but I am not very optimistic, as there are some issues which are not quite the same, but related, and they seem to be getting little attention...e.g. https://github.com/swagger-api/swagger-ui/issues/5776

### CODE REVIEW

File: app/javascript/src/ActiveDocs/components/ApiJsonSpecInput.tsx [hunk 1]
```diff
@@ -31,7 +31,7 @@ const ApiJsonSpecInput: FunctionComponent<Props> = ({
       fieldId="api_docs_service_body"
       helperText={(
         <>
-        Specification must comply with Swagger <a href="https://github.com/swagger-api/swagger-spec/blob/master/versions/1.2.md#52-api-declaration">1.2</a>, <a href="https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md">2.0</a> or <a href="https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md">3.0</a>
+        Specification must comply with Swagger <a href="https://github.com/swagger-api/swagger-spec/blob/master/versions/1.2.md#52-api-declaration">1.2</a>, <a href="https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md">2.0</a>, <a href="https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md">3.0</a> or <a href="https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.1.md">3.1</a>
```
Comment (akostadinov) [inline app/javascript/src/ActiveDocs/components/ApiJsonSpecInput.tsx:34]: If swagger 1.2 is not officially supported, perhaps state it here instead of listing it as fully supported.

File: package.json [hunk 2]
```diff
@@ -102,7 +102,7 @@
     "sass": "1.77.6",
     "sass-loader": "13.3.3",
     "swagger-client": "^3.17.0",
-    "swagger-ui": "^3.52.5",
+    "swagger-ui": "5.17.14",
```
Comment (akostadinov) [inline package.json:105]: should we hardcode the version here? Not allow minor updates?

File: app/javascript/src/ActiveDocs/components/ApiJsonSpecInput.tsx (hunk 1 above)
Comment (mayorova) [reply app/javascript/src/ActiveDocs/components/ApiJsonSpecInput.tsx:34]: Mmm, yeah, it's a good idea. I am actually not sure about its status, but makes sense to remove it from this list.

File: package.json (hunk 2 above)
Comment (mayorova) [reply package.json:105]: Yeah, I don't know... @josemigallas seemed to prefer fixed versions for npm deps lately, just to avoid unexpected issues on version bumps.
I am not sure whether it's a good idea or not.

Comment (akostadinov) [reply package.json:105]: I think that `package.json` is advisory. Once version is locked it will stay the same until updated.

So when you fix the version, I, as a reader, am thinking whether you know that future patch versions are broken or what was the reason for a fixed version.

While a normal version specification, locked to the major or minor version would indicate that we just don't want unexpected updates.

So not a blocking complaint but I think locking to the minor version is much better.

File: features/support/helpers/api_docs_service_helper.rb [hunk 3]
```diff
@@ -12,11 +13,17 @@ def fill_in_api_docs_service_body(value)
     find('.pf-c-page__main-section').click # HACK: need to click outside to lose focus
   end
 
+  def transform_swagger_version(spec_version)
+    match =  /(invalid)?\s?(Swagger 1.2|Swagger 2|OAS 3.0|OAS 3.1)/.match(spec_version)
+    { version: numbered_swagger_version(match[2]), invalid: match[1].present? }
+  end
+
   def numbered_swagger_version(version)
```
Comment (akostadinov) [inline features/support/helpers/api_docs_service_helper.rb:21]: hmm, shouldn't this method accept a named parameter `invalid` to properly work?

File: test/unit/three_scale/swagger/specification_test.rb [hunk 4]
```diff
@@ -13,6 +13,16 @@ class ThreeScale::Swagger::SpecificationTest < ActiveSupport::TestCase
     assert specification.valid?, 'specification should be valid'
   end
 
+  test "invalid 2.0 spec errors" do
+    specification = fixture_spec('invalid-2.0')
+    assert_not specification.valid?
+    errors = specification.errors.full_messages_for(:base).join(', ')
```
Comment (akostadinov) [inline test/unit/three_scale/swagger/specification_test.rb:19]: if these are separate messages that you match in full, I assume no need to `join` them. If you partially match them, this is fine though.

File: test/unit/three_scale/swagger/specification_test.rb [hunk 5]
```diff
@@ -13,6 +13,16 @@ class ThreeScale::Swagger::SpecificationTest < ActiveSupport::TestCase
...
+    assert errors.include? "The property '#/' contains additional properties [\"invalid\", \"another\"] outside of the schema when none are allowed"
```
Comment (akostadinov) [inline test/unit/three_scale/swagger/specification_test.rb:20]: we have `assert_includes` assertion that will show a more meaningful error message if it fails

File: test/unit/three_scale/swagger/specification_test.rb [hunk 6]
```diff
@@ -26,12 +36,39 @@ class ThreeScale::Swagger::SpecificationTest < ActiveSupport::TestCase
     assert specification.valid?, 'specification should be valid'
   end
 
-  test 'oas 3.0 base path' do
-    specification = ThreeScale::Swagger::Specification.new({ openapi: '3.0.0', servers: [{ url: 'https://my-petstore.io' }] }.to_json)
-    assert_equal 'https://my-petstore.io', specification.base_path
+  test 'petstore oas 3.1' do
+    specification = fixture_spec('pet-store-3.1')
+    assert specification.valid?, 'specification should be valid'
```
Comment (akostadinov) [inline test/unit/three_scale/swagger/specification_test.rb:41]: Another nitpick :)

We have `assert_valid` that might work nicely here. Not sure whether error reporting will be compatible though, has to be tested. But also we have the nice `assert_predicate specification, :valid?` that will also show a nice error message, although not as complete as the `assert_valid` would (in case it works).

File: test/unit/three_scale/swagger/specification_test.rb [hunk 7]
```diff
@@ -26,12 +36,39 @@ class ThreeScale::Swagger::SpecificationTest < ActiveSupport::TestCase
...
+
+  test 'oas 3.x base path' do
+    %w[3.0.0 3.1.0].each do |version|
+      specification = ThreeScale::Swagger::Specification.new({ openapi: version, servers: [{ url: 'https://my-petstore.io' }] }.to_json)
+      assert_equal 'https://my-petstore.io', specification.base_path
 
-    specification = ThreeScale::Swagger::Specification.new({ openapi: '3.0.0' }.to_json)
-    refute specification.base_path
+      specification = ThreeScale::Swagger::Specification.new({ openapi: version }.to_json)
+      refute specification.base_path
```
Comment (akostadinov) [inline test/unit/three_scale/swagger/specification_test.rb:70]: `assert_nil`?

File: doc/active_docs/account_management_api.json [hunk 8]
```diff
@@ -1,5 +1,5 @@
 {
-  "openapi": "3.0.1",
+  "openapi": "3.1.1",
```
[the remark this answers is not in the document]
Comment (mayorova) [reply doc/active_docs/account_management_api.json:2]: Yeah, seems that we don't use any features that were changed between these versions (nullable, exclusive minimum/maximum, examples...) https://www.openapis.org/blog/2021/02/16/migrating-from-openapi-3-0-to-3-1-0

**Review by jlledom — APPROVED** [review]

File: test/unit/three_scale/swagger/specification_test.rb (hunk 5 above)
Comment (mayorova) [reply test/unit/three_scale/swagger/specification_test.rb:20]: Yeah, I saw it, but I was kind of misled by the fact that the method definition talks about "collection" containing the object. But it's true that it can still apply here, because that assertion just calls `inlcude?` as well.

But honestly, I am not sure verbosity is good in this case.

If I introduce a change in the string, the error is currently reported as:

```
Minitest::Assertion: Expected false to be truthy.
test/unit/three_scale/swagger/specification_test.rb:20:in `block in <class:SpecificationTest>'
```
so you can easily see what the problem is - obviously, `include?` is not true.


With `assert_includes` the error shows as:

```
Minitest::Assertion: Expected "The property '#/' contains additional properties [\"invalid\", \"another\"] outside of the schema when none are allowed in schema http://swagger.io/v2/schema.json#, The property '#/info' did not contain a required property of 'title' in schema http://swagger.io/v2/schema.json#, The property '#/info' contains additional properties [\"foo\"] outside of the schema when none are allowed in schema http://swagger.io/v2/schema.json#" to include "Theq property '#/' contains additional properties [\"invalid\", \"another\"] outside of the schema when none are allowed".
test/unit/three_scale/swagger/specification_test.rb:20:in `block in <class:SpecificationTest>'
```

This is very long, and hard to follow. Obviously, if you trace back to the line, you can also see what is done there, and understand why it fails, but you would hardly read the message :grimacing:

File: test/unit/three_scale/swagger/specification_test.rb (hunk 4 above)
Comment (mayorova) [reply test/unit/three_scale/swagger/specification_test.rb:19]: Yeah, I only want to match a part of the message, because it's too long. There is an example of the full message in https://github.com/3scale/porta/pull/3934/commits/07737ed3492c0b689586b1c230739e3892eef8ae#r1839063674
I don't think we need to be that exact.

File: features/support/helpers/api_docs_service_helper.rb (hunk 3 above)
Comment (mayorova) [reply features/support/helpers/api_docs_service_helper.rb:21]: well, this method just converts e.g. `Swagger 1.2` to `1.2`. Maybe it's confusing because it looks like in the invocation both are passed? But there is a `)` after `match[2]` on like 18 :wink:

**Review by jlledom — APPROVED** [review]

**Review by lvillen — APPROVED** [review]
